src/os.rs

```diff
@@ -0,0 +1,371 @@
+    pub trait PollerKqueueExt: PollerSealed {
+        /// Add a filter to the poller.
+        ///
+        /// This is similar to [`add`][Poller::add], but it allows you to specify a filter instead of
+        /// a socket. See the implementors of the [`Filter`] trait for more information.
+        ///
+        /// # Examples
+        ///
+        /// ```no_run
+        /// use polling::{Poller, PollMode};
+        /// use polling::os::kqueue::{Filter, PollerKqueueExt, Signal};
+        ///
+        /// let poller = Poller::new().unwrap();
+        ///
+        /// // Register the SIGINT signal.
+        /// poller.add_filter(Signal::new(libc::SIGINT), 0, PollMode::Oneshot).unwrap();
+        ///
+        /// // Wait for the signal.
+        /// let mut events = vec![];
+        /// poller.wait(&mut events, None).unwrap();
+        /// # let _ = events;
+        /// ```
+        fn add_filter(&self, filter: impl Filter, key: usize, mode: PollMode) -> io::Result<()>;
+
+        /// Modify a filter in the poller.
+        ///
+        /// This is similar to [`modify`][Poller::modify], but it allows you to specify a filter
+        /// instead of a socket. See the implementors of the [`Filter`] trait for more information.
+        ///
+        /// # Examples
+        ///
+        /// ```no_run
+        /// use polling::{Poller, PollMode};
+        /// use polling::os::kqueue::{Filter, PollerKqueueExt, Signal};
+        ///
+        /// let poller = Poller::new().unwrap();
+        ///
+        /// // Register the SIGINT signal.
+        /// poller.add_filter(Signal::new(libc::SIGINT), 0, PollMode::Oneshot).unwrap();
+        ///
+        /// // Re-register with a different key.
+        /// poller.modify_filter(Signal::new(libc::SIGINT), 1, PollMode::Oneshot).unwrap();
+        ///
+        /// // Wait for the signal.
+        /// let mut events = vec![];
+        /// poller.wait(&mut events, None).unwrap();
+        /// # let _ = events;
+        /// ```
+        fn modify_filter(&self, filter: impl Filter, key: usize, mode: PollMode) -> io::Result<()>;
+
+        /// Remove a filter from the poller.
+        ///
+        /// This is used to remove filters that were previously added with
+        /// [`add_filter`](PollerKqueueExt::add_filter).
+        ///
+        /// # Examples
+        ///
+        /// ```no_run
+        /// use polling::{Poller, PollMode};
+        /// use polling::os::kqueue::{Filter, PollerKqueueExt, Signal};
+        ///
+        /// let poller = Poller::new().unwrap();
+        ///
+        /// // Register the SIGINT signal.
+        /// poller.add_filter(Signal::new(libc::SIGINT), 0, PollMode::Oneshot).unwrap();
+        ///
+        /// // Remove the filter.
+        /// poller.delete_filter(Signal::new(libc::SIGINT)).unwrap();
+        /// ```
+        fn delete_filter(&self, filter: impl Filter) -> io::Result<()>;
+    }
+
+    impl PollerSealed for Poller {}
+
+    impl PollerKqueueExt for Poller {
```

Comment:
imo it would make more sense to tie this more directly to a filter, e.g. to the tuple `(Poller, F) where F: Filter` or `impl<F: Filter> PollerKqueueExt<F> for Poller`.